File: lintcode/python/1003_binary_tree_pruning.py
#!/usr/bin/python -t

# dfs

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: the root
    @return: the same tree where every subtree (of the given tree) not containing a 1 has been removed
    """
    def pruneTree(self, root):
        # Write your code here
        if not root:
            return root
            
        root.left = self.pruneTree(root.left)
        root.right = self.pruneTree(root.right)
        
        if not root.left and not root.right and root.val == 0:
            return None
            
        return root
        
        

File: lintcode/golang/1003_binary_tree_pruning.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: the root
 * @return: the same tree where every subtree (of the given tree) not containing a 1 has been removed
 */

//import "fmt"

func pruneTree (root *TreeNode) *TreeNode {
    // Write your code here
    dpp(&root)
    return root
}

func dpp(root **TreeNode) {
    if (*root) != nil {
        dpp(&((*root).Left))
        dpp(&((*root).Right))

        if (*root).Val == 0 && (*root).Left == nil && (*root).Right == nil {
            *root = nil
        }
    }
}
